File: 4_software_and_its_engineering/computer_programming/programs_per_computational_model/0_declarative/arrays_and_strings/count_largest_object.scala
"""
  Source: https://www.hackerrank.com/challenges/birthday-cake-candles/copy-from/71884152
"""
def count_largest_object_occurrence(ar: Array[Int]): Int =
    val max_height = if (ar.length == 0) 0 else ar.reduceLeft(_ max _)
    ar.count(x => x == max_height)

@main
def main(): Unit =
    assert(count_largest_object_occurrence(Array(3, 2, 1, 3)) == 2)

    assert(
      count_largest_object_occurrence(
        Array(3, 2, 1, 3, 7, 7, 7, 1, 2, 5, 3, 3, 5, 2, 1)
      ) == 3
    )

    assert(
      count_largest_object_occurrence(Array(3, 2, 1)) == 1
    )

    assert(count_largest_object_occurrence(Array()) == 0)
    assert(count_largest_object_occurrence(Array(1)) == 1)
    assert(count_largest_object_occurrence(Array(1, 2)) == 1)

    println("completed successfully")

File: 4_software_and_its_engineering/computer_programming/programs_per_computational_model/0_declarative/arrays_and_strings/count_largest_object.py
"""
Source: https://www.hackerrank.com/challenges/birthday-cake-candles/copy-from/71884152
"""


def count_largest_object_occurrence(ar: list[int]) -> int:
    max_height = 0 if len(ar) == 0 else max(ar)
    return ar.count(max_height)


def main() -> None:
    assert count_largest_object_occurrence([3, 2, 1, 3]) == 2
    assert (
        count_largest_object_occurrence([3, 2, 1, 3, 7, 7, 7, 1, 2, 5, 3, 3, 5, 2, 1])
        == 3
    )

    assert count_largest_object_occurrence([3, 2, 1]) == 1

    assert count_largest_object_occurrence([]) == 0
    assert count_largest_object_occurrence([1]) == 1
    assert count_largest_object_occurrence([1, 2]) == 1

    print("completed successfully")


if __name__ == "__main__":
    main()
